### PR TITLE
Add base Checker classes mirroring interfaces

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Ref. #2287

### CODE REVIEW

**Comment by DanielNoord** [comment]
Failing test is due to regression in #6375.

File: pylint/checkers/base_checker.py [hunk 1]
```diff
@@ -200,16 +202,29 @@ def get_message_definition(self, msgid):
...
+    def close(self) -> None:
         """Called after visiting project (i.e set of modules)."""
 
 
 class BaseTokenChecker(BaseChecker):
     """Base class for checkers that want to have access to the token stream."""
 
-    def process_tokens(self, tokens):
+    @abc.abstractmethod
+    def process_tokens(self, tokens: list[TokenInfo]) -> None:
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base_checker.py:216]: I'd like to merge #6273 first, I let it go stale for too long, my bad.

Comment (DanielNoord) [reply pylint/checkers/base_checker.py:216]: Fine with me!

**Review by Pierre-Sassoulas — APPROVED** [review]
👌

**Pull request merged**